h/services/user_delete.py

```diff
@@ -1,70 +1,348 @@
+        """Delete all tokens belonging to `user`."""
+        self.worker.delete(Token, select(Token.id).where(Token.user == user))
+
+    def delete_flags(self, user):
+        """Delete all flags created by `user`."""
+        self.worker.delete(Flag, select(Flag.id).where(Flag.user_id == user.id))
+
+    def delete_featurecohort_memberships(self, user):
+        """Remove `user` from all feature cohorts."""
+        self.worker.delete(
+            FeatureCohortUser,
+            select(FeatureCohortUser.id).where(FeatureCohortUser.user_id == user.id),
+        )
+
+    def delete_annotations(self, user):
+        """Delete all of `user`'s annotations from both Postgres and Elasticsearch."""
+        now = datetime.utcnow()
+
+        deleted_annotation_ids = self.worker.update(
+            Annotation,
+            select(Annotation.id).where(Annotation.userid == user.userid),
+            {
+                # We don't actually delete the user's annotations from the DB,
+                # we only mark them as deleted.
+                # The marked-as-deleted annotations will later be purged by the
+                # periodic purge_deleted_annotations() task.
+                #
+                # This is because there are parts of h that don't work if
+                # annotations are deleted immediately, including the WebSocket
+                # and the call to JobQueueService.add_by_id() below.
+                #
+                # This is the same as what happens when annotations are deleted
+                # individually by the API.
+                "deleted": True,
+                # Bump the annotation's updated time when marking it as deleted.
+                # This is to prevent the purge_deleted_annotations() task from
+                # purging the annotation too soon: that task only purges
+                # deleted annotations whose updated time is more than ten mins
+                # ago.
+                "updated": now,
+            },
+        )
+
+        # Whenever we update annotations we also need to update the corresponding annotation_slims.
+        num_deleted_annotation_slims = self.db.execute(
+            update(AnnotationSlim)
+            .where(AnnotationSlim.pubid.in_(deleted_annotation_ids))
+            .values({"deleted": True, "updated": now})
+        ).rowcount
+        log.info("Updated %d rows from annotation_slim", num_deleted_annotation_slims)
+
+        # Add jobs to the queue so the annotations will eventually be deleted from Elasticsearch.
+        for annotation_id in deleted_annotation_ids:
+            self.job_queue.add_by_id(
+                name="sync_annotation",
+                annotation_id=annotation_id,
+                tag="UserDeleteService.delete_annotations",
+                schedule_in=60,
+            )
+            log.info(
+                "Enqueued job to delete annotation from Elasticsearch: %s",
+                annotation_id,
+            )
```

Comment:
This duplicates code from [`AnnotationDeleteService.delete()`](https://github.com/hypothesis/h/blob/f1669fd89ce12aa7be53cbaaf161c4dbfa81e8b8/h/services/annotation_delete.py#L23-L42) (set `annotation.deleted = True`, update `annotation.updated`, update the corresponding `annotation_slim`) but does it in a more efficient way and with support for a `limit` (via `LimitedWorker`).

This should probably be refactored to remove the duplication: `AnnotationDeleteService.delete()` should probably be refactored to work more efficiently like this `delete_annotation()` method and with support for deleting multiple annotations and using a limited worker. But I don't think I want to bite that off right now.

``AnnotationDeleteService.delete()` also [sends an `AnnotationEvent`](https://github.com/hypothesis/h/blob/f1669fd89ce12aa7be53cbaaf161c4dbfa81e8b8/h/services/annotation_delete.py#L41-L42) for the deleted annotation but I don't think I want to do that here: I don't think we want bulk-deleting annotations to cause an explosion of `AnnotationEvent`'s to be sent.